### DISCUSSION TITLE
DRY way of writing different (sub)commands with the same arguments

### DISCUSSION THREAD

**Discussion by BramVanroy** [fastapi/typer#730] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Optional

import typer
from typer import Argument, Option


app = typer.Typer()


@app.command()
def translate(
    dataset_name: Annotated[str, Argument(help="dataset name compatible with HuggingFace datasets")],
    output_directory: Annotated[str, Argument(help="output directory to save the translated dataset to")],
    config_name: Annotated[
        Optional[str],
        Option(help="optional config name for the dataset"),
    ] = None,
    **kwargs_specific_to_this_other_cmd)
    ...

@app.command()
def answer(
    dataset_name: Annotated[str, Argument(help="dataset name compatible with HuggingFace datasets")],
    output_directory: Annotated[str, Argument(help="output directory to save the translated dataset to")],
    config_name: Annotated[
        Optional[str],
        Option(help="optional config name for the dataset"),
    ] = None,
    **kwargs_specific_to_this_other_cmd)
    ...
```

### Description

I have a couple of commands that are very similar in the arguments that they take on (more than a dozen), but each of them has a few that are specific to the command. Yet, I feel like I am needlessly repeating the 95% of the signature with all the arguments/options that are identical in the two functions. Not only is this unnecessarily DRY-less, but it's also prone to errors when I forget to change something in one or the other command arguments.

Is there a way to have a base (abstract) signature command that others can inherit from? I tried functools.wrap but that did not yield the intended gain, because I still had to push kwargs through the signatures and then .get or .pop them.

I have little hope for a solution but happy to hear that I am wrong. Hoping for a simple solution.

### Operating System

Linux, Windows

### Operating System Details

OS independent

### Typer Version

0.9.0

### Python Version

3.10.10

**Comment by melsabagh** [comment, marked as answer]
How about something like: 

```python
DatasetName = Annotated[str, Argument(help="dataset name compatible with HuggingFace datasets")]
OutputDirectory = Annotated[str, Argument(help="output directory to save the translated dataset to")]
ConfigName = Annotated[
    Optional[str],
    Option(help="optional config name for the dataset"),
]


@app.command()
def translate(
    dataset_name: DatasetName,
    output_directory: OutputDirectory,
    config_name: ConfigName = None,
    **kwargs_specific_to_this_other_cmd,
): ...


@app.command()
def answer(
    dataset_name: DatasetName,
    output_directory: OutputDirectory,
    config_name: ConfigName = None,
    **kwargs_specific_to_this_other_cmd,
): ...
```

**Comment by palto42** [comment]
There are some related open issues #153, #405
